Harmonize derivative type generation over all modes

PetroZarytskyi opened the pull request:
Forward modes already have a common function ``GetPushForwardDerivativeType`` to build the derivative `QualType`. It utilizes a virtual function ``GetParameterDerivativeType`` to make the implementation consistent with the vectorized forward mode. This PR extends this approach to all modes. 
Note: function type is not the return type, it also accounts for the signature, e.g., `returnTy(parTy1, parTy2, ...)`.
There are 2 goals this PR is meant to achieve:
1) Avoid code repetition.
2) Create a common point to generate derivative QualType (derivative signature) so that it's easily available on higher levels (primarily, to perform custom derivative lookups).

PetroZarytskyi commented:
*commits 3-11 should be squashed, they are only kept separate for now for easier editing

vgvassilev commented on include/clad/Differentiator/VisitorBase.h:
```diff
@@ -133,6 +134,11 @@ namespace clad {
+    // FIXME: Should we make this an object instead of a pointer?
+    // Downside of making it an object: We will need to include
+    // 'MultiplexExternalRMVSource.h' file
+    MultiplexExternalRMVSource* m_ExternalSource = nullptr;
```
Isn't that supposed to be only for reverse mode?

vgvassilev commented:
Can we write tests for the missing coverage?

PetroZarytskyi commented:
@vgvassilev 
> Can we write tests for the missing coverage?

Addressed. The last line of ``isDifferentiableType`` should now be covered with the last commit. I removed the part handling void parameters of vector pushforwards since parameters cannot be void and the function is never used for anything else.

vgvassilev reviewed (approved):
LGTM. Consider addressing clang-format reports if they are applicable and squash on merging.
